File: src/Butil/Bit.Butil/Scripts/webTransport.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    interface Session {
        transport: any;
        // Held per session so a later openStream call doesn't have to be handed the .NET reference
        // again - the streams it opens dispatch their data through the same one.
        dotNetRef: any;
        streams: { [streamId: string]: { writer: any; readable: any } };
        nextStream: number;
        closed: boolean;
    }

    const _sessions: { [id: string]: Session } = {};

    function toBytes(value: any): Uint8Array {
        if (value instanceof Uint8Array) return value;
        return new Uint8Array(value ?? []);
    }

    // A reader loop per stream, per session. Every WebTransport read is a pull, so the only way to
    // surface incoming data to .NET is to keep pulling and dispatch each chunk as it arrives.
    async function pump(dotNetRef: any, id: string, streamId: string, readable: any) {
        const reader = readable.getReader();
        try {
            for (; ;) {
                const { value, done } = await reader.read();
                if (done) break;
                butil.utils.dispatch(dotNetRef, 'InvokeWebTransportStreamData', id, streamId, toBytes(value));
            }
        } catch {
            // The stream was reset or the session went away; the closed callback reports the why.
        } finally {
            try { reader.releaseLock(); } catch { /* already released */ }

            // The read side is finished for good, so a stream with nothing left to write on is
            // dropped here - otherwise a long-lived session grows its map by one entry for every
            // unidirectional stream the server ever opened. One that still has a writer stays
            // until closeStream/close, since .NET can go on answering on it after the peer has
            // finished sending. The identity check keeps a later stream reusing the id safe.
            const stream = _sessions[id]?.streams[streamId];
            if (stream && stream.readable === readable && !stream.writer) delete _sessions[id].streams[streamId];

            butil.utils.dispatch(dotNetRef, 'InvokeWebTransportStreamEnd', id, streamId);
        }
    }

    async function pumpDatagrams(dotNetRef: any, id: string, transport: any) {
        const reader = transport.datagrams?.readable?.getReader?.();
        if (!reader) return;
        try {
            for (; ;) {
                const { value, done } = await reader.read();
                if (done) break;
                butil.utils.dispatch(dotNetRef, 'InvokeWebTransportDatagram', id, toBytes(value));
            }
        } catch {
            // Datagrams stop with the session.
        }
    }

    async function pumpIncomingStreams(dotNetRef: any, id: string, session: Session, source: any, bidirectional: boolean) {
        if (!source) return;
        const reader = source.getReader();
        try {
            for (; ;) {
                const { value, done } = await reader.read();
                if (done) break;

                const streamId = `s${session.nextStream++}`;
                // A unidirectional stream from the server IS the readable; a bidirectional one
                // carries both halves, and the writable is what lets .NET answer on it.
                const readable = bidirectional ? value.readable : value;
                session.streams[streamId] = { writer: bidirectional ? value.writable.getWriter() : null, readable };
                butil.utils.dispatch(dotNetRef, 'InvokeWebTransportStreamOpened', id, streamId, bidirectional);
                pump(dotNetRef, id, streamId, readable);
            }
        } catch {
            // The session ended.
        }
    }

    butil.webTransport = {
        isSupported() { return typeof (window as any).WebTransport === 'function'; },
        async connect(dotNetRef: any, id: string, url: string, allowPooling: boolean, congestionControl: string, certificateHashes: any[]) {
            const WT = (window as any).WebTransport;
            if (typeof WT !== 'function') return { connected: false, error: 'WebTransport is not supported' };

            const options: any = {};
            if (allowPooling) options.allowPooling = true;
            if (congestionControl) options.congestionControl = congestionControl;
            if (certificateHashes?.length) {
                // Only a serverCertificateHashes connection can talk to a server whose certificate
                // no public CA signed - the usual case for a local HTTP/3 endpoint.
                options.serverCertificateHashes = certificateHashes.map(hash => ({
                    algorithm: hash.algorithm || 'sha-256',
                    value: toBytes(hash.value)
                }));
            }

            let transport: any;
            try {
                transport = new WT(url, options);
            } catch (e: any) {
                // A url that is not https, or a malformed one - the only synchronous failure.
                return { connected: false, error: String(e?.message ?? e) };
            }

            const session: Session = { transport, dotNetRef, streams: {}, nextStream: 1, closed: false };
            _sessions[id] = session;

            try {
                await transport.ready;
            } catch (e: any) {
                if (_sessions[id] === session) delete _sessions[id];
                // transport.closed rejects alongside ready for a connection that never came up.
                // Nothing observes it yet - the closed callback is only attached below, once the
                // session is established - so swallow it here rather than leave it unhandled.
                try { transport.closed?.catch?.(() => { }); } catch { /* no closed promise */ }
                return { connected: false, error: String(e?.message ?? e) };
            }

            // Attached only after ready resolved, so onClosed reports the end of a session that
            // existed and never doubles as a failed-connect callback - Connect's own return value
            // is what reports that. A session the peer or the network ended is never closed through
            // close(), so this is also the only place it leaves _sessions - guarded on identity,
            // since the same id could already have been replaced by a later connect.
            const forget = () => { if (_sessions[id] === session) delete _sessions[id]; };

            transport.closed
                .then((closeInfo: any) => {
                    session.closed = true;
                    butil.utils.dispatch(dotNetRef, 'InvokeWebTransportClosed', id, closeInfo?.closeCode ?? 0, closeInfo?.reason ?? '', '');
                    forget();
                })
                .catch((e: any) => {
                    session.closed = true;
                    butil.utils.dispatch(dotNetRef, 'InvokeWebTransportClosed', id, 0, '', String(e?.message ?? e));
                    forget();
                });

            pumpDatagrams(dotNetRef, id, transport);
            pumpIncomingStreams(dotNetRef, id, session, transport.incomingUnidirectionalStreams, false);
            pumpIncomingStreams(dotNetRef, id, session, transport.incomingBidirectionalStreams, true);

            return { connected: true, error: '' };
        },
        async sendDatagram(id: string, data: Uint8Array) {
            const session = _sessions[id];
            if (!session || session.closed) return false;
            try {
                const writer = session.transport.datagrams.writable.getWriter();
                try { await writer.write(data); } finally { writer.releaseLock(); }
                return true;
            } catch {
                // Datagrams are unreliable by design: one too large for the path MTU, or sent into
                // a full queue, is dropped rather than delivered.
                return false;
            }
        },
        async openStream(id: string, bidirectional: boolean) {
            const session = _sessions[id];
            if (!session || session.closed) return null;
            try {
                const streamId = `s${session.nextStream++}`;
                if (bidirectional) {
                    const stream = await session.transport.createBidirectionalStream();
                    session.streams[streamId] = { writer: stream.writable.getWriter(), readable: stream.readable };
                    pump(session.dotNetRef, id, streamId, stream.readable);
                } else {
                    const writable = await session.transport.createUnidirectionalStream();
                    session.streams[streamId] = { writer: writable.getWriter(), readable: null };
                }
                return streamId;
            } catch {
                return null;
            }
        },
        async writeStream(id: string, streamId: string, data: Uint8Array) {
            const stream = _sessions[id]?.streams[streamId];
            if (!stream?.writer) return false;
            try { await stream.writer.write(data); return true; } catch { return false; }
        },
        async closeStream(id: string, streamId: string) {
            const session = _sessions[id];
            const stream = session?.streams[streamId];
            if (!stream) return false;
            delete session.streams[streamId];
            try { await stream.writer?.close(); } catch { /* already closed or reset */ }
            return true;
        },
        state(id: string) {
            const session = _sessions[id];
            if (!session) return 'closed';
            return session.closed ? 'closed' : 'open';
        },
        async close(id: string, closeCode: number, reason: string) {
            const session = _sessions[id];
            if (!session) return;
            delete _sessions[id];
            session.closed = true;
            for (const streamId of Object.keys(session.streams)) {
                try { await session.streams[streamId].writer?.close(); } catch { /* already gone */ }
            }
            try { session.transport.close({ closeCode, reason }); } catch { /* already closed */ }
        },
        async disposeAll() {
            for (const id of Object.keys(_sessions)) await butil.webTransport.close(id, 0, '');
        }
    };
}(BitButil));

File: src/Butil/Bit.Butil/Publics/WebTransport.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/WebTransport_API">WebTransport API</see>:
/// a bidirectional connection to an HTTP/3 server carrying many independent streams and, separately,
/// unreliable datagrams.
/// </summary>
/// <remarks>
/// What it has over a WebSocket is head-of-line blocking, or the lack of it. A WebSocket is one
/// ordered stream over TCP, so a lost packet stalls everything behind it; WebTransport runs over
/// QUIC, where each stream is independent and a datagram is not retransmitted at all. That matters
/// for games, live media and telemetry, and matters not at all for a chat message.
/// <br/>
/// Requires an HTTP/3 server: this cannot talk to an ordinary HTTPS endpoint, and there is no
/// fallback to one. For a development server whose certificate no public CA signed, pass its hash
/// through <c>certificateHashes</c>.
/// <br/>
/// Incoming data is delivered through the callbacks passed to <see cref="Connect"/> rather than
/// read from the handle: the browser's reader is a pull loop, so it runs on the JS side and
/// dispatches what it reads.
/// </remarks>
[ButilService(typeof(WebTransport))]
public class WebTransport(IJSRuntime js) : IAsyncDisposable
{
    internal const string DatagramMethodName = nameof(InvokeWebTransportDatagram);
    internal const string StreamDataMethodName = nameof(InvokeWebTransportStreamData);
    internal const string StreamEndMethodName = nameof(InvokeWebTransportStreamEnd);
    internal const string StreamOpenedMethodName = nameof(InvokeWebTransportStreamOpened);
    internal const string ClosedMethodName = nameof(InvokeWebTransportClosed);

    private readonly ConcurrentDictionary<Guid, SessionHandlers> _handlers = new();

    // Per-instance callback reference (see Keyboard): sessions are isolated per circuit / WASM app
    // and released on disposal - no static state, no cross-circuit leak.
    private DotNetObjectReference<WebTransport>? _dotNetRef;
    private DotNetObjectReference<WebTransport> DotNetRef => DotNetObjectReferenceHelper.GetOrCreate(ref _dotNetRef, this);

    private sealed record SessionHandlers(
        Action<byte[]>? OnDatagram,
        Action<WebTransportStreamData>? OnStreamData,
        Action<WebTransportStream>? OnStreamOpened,
        Action<WebTransportCloseInfo>? OnClosed);

    /// <summary>True when the runtime exposes <c>WebTransport</c>.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.webTransport.isSupported");

    /// <summary>
    /// Invoked from JS for each datagram. Public + <see cref="JSInvokableAttribute"/> so it can be
    /// dispatched through the per-instance <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    [JSInvokable(DatagramMethodName)]
    public void InvokeWebTransportDatagram(Guid id, byte[] data)
    {
        if (_handlers.TryGetValue(id, out var handlers)) handlers.OnDatagram?.Invoke(data);
    }

    /// <summary>Invoked from JS for each chunk read off a stream. See <see cref="InvokeWebTransportDatagram"/>.</summary>
    [JSInvokable(StreamDataMethodName)]
    public void InvokeWebTransportStreamData(Guid id, string streamId, byte[] data)
    {
        if (_handlers.TryGetValue(id, out var handlers))
            handlers.OnStreamData?.Invoke(new WebTransportStreamData(streamId, data, false));
    }

    /// <summary>Invoked from JS when a stream is finished. See <see cref="InvokeWebTransportDatagram"/>.</summary>
    [JSInvokable(StreamEndMethodName)]
    public void InvokeWebTransportStreamEnd(Guid id, string streamId)
    {
        if (_handlers.TryGetValue(id, out var handlers))
            handlers.OnStreamData?.Invoke(new WebTransportStreamData(streamId, [], true));
    }

    /// <summary>Invoked from JS when the server opens a stream. See <see cref="InvokeWebTransportDatagram"/>.</summary>
    [JSInvokable(StreamOpenedMethodName)]
    public void InvokeWebTransportStreamOpened(Guid id, string streamId, bool bidirectional)
    {
        if (_handlers.TryGetValue(id, out var handlers))
            handlers.OnStreamOpened?.Invoke(new WebTransportStream(js, id, streamId, bidirectional));
    }

    /// <summary>Invoked from JS when the session ends. See <see cref="InvokeWebTransportDatagram"/>.</summary>
    /// <remarks>
    /// This is where a session's handlers are dropped, rather than when <see cref="WebTransportHandle.Close"/>
    /// was called: the closure is dispatched after that call, so removing them any earlier would
    /// swallow the notification for every session this side closed.
    /// </remarks>
    [JSInvokable(ClosedMethodName)]
    public void InvokeWebTransportClosed(Guid id, int closeCode, string reason, string error)
    {
        if (_handlers.TryRemove(id, out var handlers))
            handlers.OnClosed?.Invoke(new WebTransportCloseInfo(closeCode, reason, error));
    }

    /// <summary>
    /// Connects to an HTTP/3 endpoint and waits until the session is usable.
    /// </summary>
    /// <param name="url">An <c>https:</c> URL served by a WebTransport-capable HTTP/3 server.</param>
    /// <param name="onDatagram">Called for each datagram received.</param>
    /// <param name="onStreamData">
    /// Called for each chunk read off any stream of this session, and once more with
    /// <see cref="WebTransportStreamData.Ended"/> when a stream finishes.
    /// </param>
    /// <param name="onStreamOpened">
    /// Called when the <em>server</em> opens a stream. The stream passed in is how to answer on it
    /// when it is bidirectional; streams this side opens come back from
    /// <see cref="WebTransportHandle.OpenStream"/> instead.
    /// </param>
    /// <param name="onClosed">
    /// Called once when the session ends, whichever side ended it and whether it closed or failed.
    /// Not called for a connection that never established - that is this method's own result.
    /// </param>
    /// <param name="allowPooling">Allow reusing an existing QUIC connection to the same host instead of opening a new one.</param>
    /// <param name="congestionControl">What to tune the connection for.</param>
    /// <param name="certificateHashes">
    /// Accept exactly these server certificates rather than the ones a public CA signed - how to
    /// reach a development server. See <see cref="WebTransportCertificateHash"/> for its limits.
    /// </param>
    /// <returns>
    /// The session, or the reason there isn't one. See <see cref="WebTransportConnectResult"/>.
    /// </returns>
    /// <remarks>
    /// Dispose the returned session when you're done; an undisposed one holds the connection open
    /// for as long as the page lives.
    /// </remarks>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(WebTransportCertificateHash))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(WebTransportConnectInfo))]
    [DynamicDependency(nameof(InvokeWebTransportDatagram), typeof(WebTransport))]
    [DynamicDependency(nameof(InvokeWebTransportStreamData), typeof(WebTransport))]
    [DynamicDependency(nameof(InvokeWebTransportStreamEnd), typeof(WebTransport))]
    [DynamicDependency(nameof(InvokeWebTransportStreamOpened), typeof(WebTransport))]
    [DynamicDependency(nameof(InvokeWebTransportClosed), typeof(WebTransport))]
    public async ValueTask<WebTransportConnectResult> Connect(string url,
                                                              Action<byte[]>? onDatagram = null,
                                                              Action<WebTransportStreamData>? onStreamData = null,
                                                              Action<WebTransportStream>? onStreamOpened = null,
                                                              Action<WebTransportCloseInfo>? onClosed = null,
                                                              bool allowPooling = false,
                                                              WebTransportCongestionControl congestionControl = WebTransportCongestionControl.Default,
                                                              WebTransportCertificateHash[]? certificateHashes = null)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(url);

        var id = Guid.NewGuid();
        _handlers[id] = new SessionHandlers(onDatagram, onStreamData, onStreamOpened, onClosed);

        var info = await js.Invoke<WebTransportConnectInfo?>("BitButil.webTransport.connect",
            DotNetRef, id, url, allowPooling, CongestionControl(congestionControl), certificateHashes ?? []);

        if (info?.Connected is not true)
        {
            _handlers.TryRemove(id, out _);
            return new WebTransportConnectResult(null, info?.Error ?? "the connection could not be established");
        }

        return new WebTransportConnectResult(
            new WebTransportHandle(js, id, () => _handlers.TryRemove(id, out _)),
            string.Empty);
    }

    /// <summary>
    /// On scope/circuit teardown, closes any session whose <see cref="WebTransportHandle"/> was
    /// never disposed.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            _handlers.Clear();
            await js.InvokeVoid("BitButil.webTransport.disposeAll");
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed
        finally
        {
            _dotNetRef?.Dispose();
            _dotNetRef = null;
        }
        GC.SuppressFinalize(this);
    }

    private static string? CongestionControl(WebTransportCongestionControl congestionControl) => congestionControl switch
    {
        WebTransportCongestionControl.Throughput => "throughput",
        WebTransportCongestionControl.LowLatency => "low-latency",
        _ => null,
    };
}
